module/zfs/zfs_znode.c

```diff
@@ -1357,6 +1361,35 @@ zfs_compare_timespec(struct timespec *t1, struct timespec *t2)
 	return (t1->tv_nsec - t2->tv_nsec);
 }
 
+/*
+ * Determine whether the znode's atime must be updated.  The logic mostly
+ * duplicates the Linux kernel's relatime_need_update() functionality.
+ * This function is only called if the underlying filesystem actually has
+ * atime updates enabled.
+ */
+boolean_t
+zfs_relatime_need_update(const struct inode *ip)
+{
+	inode_timespec_t now;
+
+	gethrestime(&now);
+	/*
+	 * In relatime mode, only update the atime if the previous atime
+	 * is earlier than either the ctime or mtime or if at least a day
+	 * has passed since the last update of atime.
+	 */
+	if (zfs_compare_timespec(&ip->i_mtime, &ip->i_atime) >= 0)
+		return (B_TRUE);
+
+	if (zfs_compare_timespec(&ip->i_ctime, &ip->i_atime) >= 0)
+		return (B_TRUE);
+
+	if ((long)now.tv_sec - ip->i_atime.tv_sec >= 24*60*60)
```

Comment:
nit: to make sure this is correct on all architectures can you cast both `now.tv_sec` and `ip->i_atime.tv_sec` to the `hrtime_t` type.  This will always be a 64-bit signed type.

Reply:
fixed